encoding/protobuf.go

```diff
@@ -112,19 +113,42 @@ func (p protoSerializer) Encoding() Encoding {
 	return Protobuf
 }
 
+func (p protoSerializer) IsClientStreaming() bool {
+	return p.method.IsClientStreaming()
+}
+
+func (p protoSerializer) IsServerStreaming() bool {
+	return p.method.IsServerStreaming()
+}
+
+func (p protoSerializer) StreamRequest() ([]byte, error) {
+	if !p.IsClientStreaming() && !p.IsServerStreaming() {
+		return nil, errors.New("proto method does not support streaming")
+	}
+	jsonBytes, err := p.reqDecoder.Next()
+	if err != nil {
+		return nil, err
+	}
+	return p.encode(jsonBytes)
+}
+
 func (p protoSerializer) Request() (*transport.Request, error) {
+	if p.IsClientStreaming() || p.IsServerStreaming() {
+		return &transport.Request{
+			Method: procedure.ToName(p.serviceName, p.methodName),
+		}, nil
+	}
+
```

Comment:
This is a review not only for this PR but for https://github.com/yarpc/yab/pull/314, https://github.com/yarpc/yab/pull/316 and https://github.com/yarpc/yab/pull/317. This change is the cause of everything :)

What is the real reason of changing `Request(input []byte) (*transport.Request, error)` to `Request() (*transport.Request, error)` ? If we create a new method `StreamRequest(io.Reader)` and `NextStreamRequestBody()` would we need all of the changes with `Request` and the constructor?
With this proposal we are in a weird state where `Request()` does a lot of things that nothing the API indicates (encoding, special case for streaming)

Reply:
Below a solution we could have to avoid to break APIs and have consistent/logic APIs naming:

First, we could create a new StreamRequestReader for the stream request:

```
type StreamRequestReader interface {
	NextBody() ([]byte, error)
}
type protoStreamerSerializer struct {
	reader io.Reader
	s protoSerializer
}
func (p protoStreamerSerializer)NextBody() (body []byte, error){
        // private encode method to encode a grpc body
	return p.s.encode(p.reader.Next())
}

func (p protoSerializer) StreamRequest(body io.Reader) (*transport.Request, StreamRequestReader, error) {
....
}
```

Serialize and StreamSerializer  interface:
```
// Serializer serializes and deserializes data for a specific encoding and method.
type Serializer interface {
	// Encoding returns the encoding for this serializer.
	Encoding() Encoding
	// Request creates a transport.Request from the given []byte input.
        // returns an error if reader has several entries
	Request(body io.Reader) (*transport.Request, error)
	// Response converts a transport.Response into something that can be displayed to a user.
	// For non-raw encodings, this is typically a map[string]interface{}.
	Response(body *transport.Response) (response interface{}, err error)
	// CheckSuccess checks whether the response body is a success, and if not, returns an
	// error with the failure reason.
	CheckSuccess(body *transport.Response) error
}

type StreamSerializer interface {
   StreamRequest(body io.Reader) (*transport.Request, StreamRequestReader, error)
}
```
Here we have `StreamRequest` and `Request` which both take a io.reader param (so it is consistent in the API).

What I would also recommend here:
- Request should call a private method `encode` to encode the body
- Request return an error if being used with a streaming API
- Request could return an error if several entries in the io.reader? (I am still thinking of the value of this thought, what do you think?)

Thoughts?

Reply:
This looks great, something on the lines of how I am refactoring it currently

>Request could return an error if several entries in the io.reader? (I am still thinking of the value of this thought, what do you think?)

I would avoid doing any additional input decoding to find if there are more than one request since this issue exists currently, will fix this in a separate PR

Reply:
I haven't modified `serializer.Request([]byte)` to use `body io.Reader` in this PR, this is to keep the change very specific to streaming methods only.

IMO, since unary requests have single request having a body io.Reader might be an overkill. This method could continue expecting the bytes from caller.